### PR TITLE
Add Splunk CIM fields for web

### LINKED ISSUE #38: Map Heroku request logs to the Splunk CIM for Web
See https://docs.splunk.com/Documentation/CIM/5.0.1/User/Web.

I believe we would use field aliases to align the fields sent by Heroku with the CIM fields for Web.

This would benefit the Enterprise Security product which leans on the Splunk common information models. It would enable the inclusion of Heroku request logs in any evaluations that the SEIM product runs.

### PR DESCRIPTION
**Pull request by sjparkinson**
Based on https://docs.splunk.com/Documentation/CIM/5.0.1/User/Web.

Resolves #38.

**Screenshots**

image

Screenshot 2022-08-03 at 08 42 21

### CODE REVIEW

File: app/default/props.conf [hunk 1]
```diff
@@ -19,24 +19,31 @@ TRANSFORMS-source_type = heroku_system_source_type, heroku_requests_source_type,
 # Do not index some high volume low value messages.
 TRANSFORMS-heroku_requests_null_queue = heroku_requests_null_queue
 
-[heroku_system]
+[heroku:system]
```
Comment (sjparkinson) [inline app/default/props.conf:22]: Renaming the source types to use a `:` convention, based on the tips in https://www.splunk.com/en_us/blog/tips-and-tricks/sourcetypes-whats-in-name.html.

File: app/default/props.conf [hunk 2]
```diff
@@ -19,24 +19,37 @@ TRANSFORMS-source_type = heroku_system_source_type, heroku_requests_source_type,
...
+
+[heroku:api]
 category = Structured
 description = Messages about Heroku app administrative actions taken by users.
 
-EXTRACT-heroku_component,dyno = ^\S+\s(?<component>\S+)\s(?<dyno>\S+)
+EXTRACT-component,dyno = ^\S+\s(?<component>\S+)\s(?<dyno>\S+)
+
+# Splunk CIM fields for the change model, see https://docs.splunk.com/Documentation/CIM/5.0.1/User/Change.
+EXTRACT-user = by user (?<user>\S+)
```
Comment (sjparkinson) [inline app/default/props.conf:55]: An initial field extraction for the Change model.

There is lots of good stuff in this one so I'll raise an issue to capture what we could do with these messages.

File: app/default/props.conf (hunk 1 above)
Comment (sjparkinson) [reply app/default/props.conf:22]: I'd also like to explore renaming `heroku` to `heroku:app` in another pull request. Keep `heroku` as the "source type router" and add a new one for `heroku:app`, might be possible but need to test it.

**Comment by sjparkinson** [comment]
@rowanmanning @alexmuller I'd be interested in your feedback on this change, mainly on the changes to source type names.

The CIM fields make this app compatible with the Splunk Enterprise Security product, I'll work with Cyber Security to integrate it.

**Comment by sjparkinson** [comment]
Note: valid values for `fwd` include commas and more than one IP address, e.g. `fwd="10.0.0.0, 12.0.0.0,14.0.0.0"`...

File: app/default/props.conf (hunk 1 above)
Comment (rowanmanning) [reply app/default/props.conf:22]: I like semi-colons in the source types 👍 if we're making this change (and the `heroku` -> `heroku:app` one) then it'd be great to do so before we roll out to too many Customer Products repos. I'd rather not have to update migration guides and dashboards too many times

Comment (sjparkinson) [reply app/default/props.conf:22]: Yep I see this as the final get naming right before the big adoption is there 😂

**Comment by sjparkinson** [comment]
@mkk19 would you be up for reviewing this one? Happy to run through the changes together!

File: app/default/props.conf (hunk 1 above)
Comment (sjparkinson) [reply app/default/props.conf]: Split out the source type renaming into #41.

File: README.md [hunk 3]
```diff
@@ -1,7 +1,18 @@
...
+This Splunk app is compatible with the [Splunk Common Information Model](https://docs.splunk.com/Documentation/CIM/5.0.1/User/Overview) for [Web](https://docs.splunk.com/Documentation/CIM/5.0.1/User/Web) and [Change](https://docs.splunk.com/Documentation/CIM/5.0.1/User/Change).
+
+## Source types
+
+| Source type     | Description                                                                                               |
+|-----------------|-----------------------------------------------------------------------------------------------------------|
+| `heroku`        | Logging output from the application itself, including logs generated by your app’s code and dependencies. |
+| `heroku:router` | Request logs produced by the Heroku HTTP routers.                                                         |
+| `heroku:system` | Messages about actions taken by the Heroku platform.                                                      |
+| `heroku:api`    | Messages about Heroku app administrative actions taken by users.                                          |
```
Comment (sjparkinson) [inline README.md:14]: I'll get this corrected in #41

File: app/default/props.conf [hunk 4]
```diff
@@ -33,10 +33,26 @@ category = Structured
...
+# Splunk CIM fields for the web model, see https://docs.splunk.com/Documentation/CIM/5.0.1/User/Web and https://devcenter.heroku.com/articles/http-routing#info-logs.
+# The `status` field already exists in the log message.
+EXTRACT-duration = service=(?<duration>\d+)ms
+EXTRACT-uri_path = path="(?<uri_path>[^#?"]+)\S*"
+EXTRACT-uri_query = path="\S+(?<uri_query>\?[^"]+)"
+EVAL-src = mvindex(fwd, 0)
+EVAL-url = protocol . "://" . host . path
+FIELDALIAS-http_method = method AS http_method
+FIELDALIAS-dest = host AS dest
+FIELDALIAS-bytes_out = bytes AS bytes_out
```
Comment (sjparkinson) [inline app/default/props.conf:49]: Using `FIELDALIAS` to copy one existing field in to a new field with with a different name, the one the CIM is expecting!

File: app/default/props.conf [hunk 5]
```diff
@@ -33,10 +33,26 @@ category = Structured
...
+EXTRACT-component = ^\S+\s(?<component>\S+)
+
+# We convert `fwd` into a multivalue field using `fields.conf`.
+
+# Splunk CIM fields for the web model, see https://docs.splunk.com/Documentation/CIM/5.0.1/User/Web and https://devcenter.heroku.com/articles/http-routing#info-logs.
+# The `status` field already exists in the log message.
+EXTRACT-duration = service=(?<duration>\d+)ms
+EXTRACT-uri_path = path="(?<uri_path>[^#?"]+)\S*"
+EXTRACT-uri_query = path="\S+(?<uri_query>\?[^"]+)"
+EVAL-src = mvindex(fwd, 0)
```
Comment (sjparkinson) [inline app/default/props.conf:45]: Take the first value from the multivalue `fwd` field. These IP addresses come from the [`X-Forwarded-For` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For). The left most IP _should_ be the client IP address.

File: app/default/tags.conf [hunk 6]
```diff
@@ -0,0 +1,3 @@
+[sourcetype=heroku_requests]
+web = enabled
+proxy = enabled
```
Comment (sjparkinson) [inline app/default/tags.conf:3]: This marks our heroku request messages as compatible with the web CIM. As a result they should be available to view in the Web Dataset in Splunk once the indexes are added to the `cim_Web_indexes` macro.

**Review by Aftab-tech — APPROVED** [review]

**Pull request merged**